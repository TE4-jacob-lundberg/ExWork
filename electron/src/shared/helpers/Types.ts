export interface ILink {
  url: string;
  label: string;
}

export interface IGameFormFields {
  id: string;
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  links: ILink[];
  fileNames: string[];
}
