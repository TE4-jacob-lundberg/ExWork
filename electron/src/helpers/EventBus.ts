/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
