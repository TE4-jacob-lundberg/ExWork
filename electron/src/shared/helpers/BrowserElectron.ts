// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
