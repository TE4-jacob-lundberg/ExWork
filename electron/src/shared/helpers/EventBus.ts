/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
