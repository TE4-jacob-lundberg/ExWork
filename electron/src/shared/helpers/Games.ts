import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
