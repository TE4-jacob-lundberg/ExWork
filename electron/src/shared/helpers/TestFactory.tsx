/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
