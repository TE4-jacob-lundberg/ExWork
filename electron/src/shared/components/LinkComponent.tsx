import React from 'react';
import Styled from '@emotion/styled';

import { ILink } from '../helpers/Types';
import { electron } from '../helpers/BrowserElectron';

const ipcRenderer = electron.ipcRenderer;

interface Props {
  link: ILink
  background?: string;
}

const LinkComponent: React.FC<Props> = function (props: Props) {
  function handleClick(): void {
    ipcRenderer.invoke('open-link', props.link.url)
      .catch((err: ErrorEvent) => console.error(err));
  }

  const ContainerStyled = Styled.div`
    min-height: 50px;
    display: flex;
    align-items: center;
    width: 100%;
    background: ${props.background};

    &:hover { 
      background: white;
    }
  `;
  
  const ButtonStyled = Styled.button`
    background: none;
    border: none;
    width: 100%;
    color: #FFF;
    height: 100%;
    font-size: 1.25rem;
    padding: 12px 0;

    &:hover {
      color: #000;
    }
  `;
  
  return (
    <ContainerStyled>
      <ButtonStyled onClick={handleClick}>
        {props.link.label}
      </ButtonStyled>  
    </ContainerStyled>
  );
};

export default LinkComponent;
