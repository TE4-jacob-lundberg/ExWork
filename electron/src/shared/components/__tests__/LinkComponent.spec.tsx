import React from 'react';

import { factory } from '../../helpers/TestFactory';
import LinkComponent from '../LinkComponent';

jest.mock('../../helpers/BrowserElectron.ts');

const testComponent = <LinkComponent 
  link={{label: 'test', url: 'test.com'}}
/>;

describe('LinkComponent.tsx', () => {
  it('renders as expected', () => {
    const { container } = factory(testComponent);
    
    expect(container).toMatchSnapshot();
  });
});
