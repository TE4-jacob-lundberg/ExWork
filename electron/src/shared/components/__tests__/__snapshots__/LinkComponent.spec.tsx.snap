// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkComponent.tsx renders as expected 1`] = `
<div>
  <div
    class="css-cfdpzz"
  >
    <i
      class="material-icons css-dyscyh"
    >
      remove_circle
    </i>
    <button
      class="css-7kw57i"
    >
      test
    </button>
  </div>
</div>
`;
